### PR TITLE
Add output settings

### PR DESCRIPTION
**Pull request by Dami-star**

### CODE REVIEW

File: examples/tinywl-new/OutputMenuBar.qml [hunk 1]
```diff
@@ -8,6 +8,7 @@ import Tinywl
 
 Item {
     required property PrimaryOutput output
+    property bool modeSwitch: false
```
Comment (zccrs) [inline examples/tinywl-new/OutputMenuBar.qml:11]: Get mode's value from `Helper`

File: examples/tinywl-new/helper.cpp [hunk 2]
```diff
@@ -644,6 +644,65 @@ Output *Helper::getOutput(WOutput *output) const
...
+
+void Helper::setShowMode(ShowMode mode)
+{
+    if (mode == ShowMode::Copy) {
+        if (m_outputList.length() < 2)
+            return;
+
+        for (int i = 1; i < m_outputList.size();) {
+            Output *o = Output::createCopy(m_outputList.at(i)->output(), m_outputList.at(0), qmlEngine(), this);
+            WOutputViewport *viewportPrimary = m_outputList.at(0)->outputItem()->findChild<WOutputViewport*>({}, Qt::FindDirectChildrenOnly);
```
Comment (zccrs) [inline examples/tinywl-new/helper.cpp:670]: Use `RootSurfaceContainer::primaryOutput`

File: examples/tinywl-new/helper.cpp [hunk 3]
```diff
@@ -644,6 +644,65 @@ Output *Helper::getOutput(WOutput *output) const
...
+
+            m_surfaceContainer->addOutput(o);
+            enableOutput(o->output());
+
+            output->deleteLater();
+        }
+
+        m_outputCopyList.clear();
+    }
+
```
Comment (zccrs) [inline examples/tinywl-new/helper.cpp:703]: remove

File: examples/tinywl-new/helper.cpp [hunk 4]
```diff
@@ -644,6 +644,65 @@ Output *Helper::getOutput(WOutput *output) const
...
+            Output *o = Output::createCopy(m_outputList.at(i)->output(), m_outputList.at(0), qmlEngine(), this);
+            WOutputViewport *viewportPrimary = m_outputList.at(0)->outputItem()->findChild<WOutputViewport*>({}, Qt::FindDirectChildrenOnly);
+            WOutputViewport *viewportCopy = o->outputItem()->findChild<WOutputViewport*>({}, Qt::FindDirectChildrenOnly);
+            Q_ASSERT(viewportCopy);
+            auto textureProxy = o->outputItem()->findChild<WQuickTextureProxy*>();
+            Q_ASSERT(textureProxy);
+
+            m_outputCopyList.append(o);
+
+            QList<WOutputLayer *>  m_hardwareLayersOfPrimaryOutput = viewportPrimary->hardwareLayers();
```
Comment (zccrs) [inline examples/tinywl-new/helper.cpp:678]: Watch hardwareLayersChanged

File: examples/tinywl-new/helper.cpp [hunk 5]
```diff
@@ -644,6 +644,65 @@ Output *Helper::getOutput(WOutput *output) const
...
+            for (auto layer : std::as_const(m_hardwareLayersOfPrimaryOutput))
+                m_renderWindow->attach(layer, viewportCopy, viewportPrimary, textureProxy);
```
Comment (zccrs) [inline examples/tinywl-new/helper.cpp:680]: Can do in `Output::createCopy`?

File: examples/tinywl-new/output.cpp [hunk 6]
```diff
@@ -50,7 +53,13 @@ Output *Output::createPrimary(WOutput *output, QQmlEngine *engine, QObject *pare
 Output *Output::createCopy(WOutput *output, Output *proxy, QQmlEngine *engine, QObject *parent)
 {
     QQmlComponent delegate(engine, "Tinywl", "CopyOutput");
-    QObject *obj = delegate.create(engine->rootContext());
+
+    WOutputViewport *viewport = proxy->outputItem()->findChild<WOutputViewport*>({}, Qt::FindDirectChildrenOnly);
```
Comment (zccrs) [inline examples/tinywl-new/output.cpp:57]: Needs use `onscreenViewport` object, the outputviewport maybe changed after create copy output.

File: examples/tinywl-new/rootsurfacecontainer.cpp [hunk 7]
```diff
@@ -44,6 +44,11 @@ RootSurfaceContainer::RootSurfaceContainer(QQuickItem *parent)
     });
 
     connect(m_outputLayout, &WOutputLayout::notify_change, this, [this] {
+
+        // The default layout is horizontal to the right
+        if (m_currentOutput && m_outputLayout->implicitWidth())
+            m_currentOutput->outputItem()->setX(m_outputLayout->implicitWidth());
```
Comment (zccrs) [inline examples/tinywl-new/rootsurfacecontainer.cpp:50]: Must update position for which output:

````
for (....)
    output->updatePositionFromOutputLayout();
````

File: examples/tinywl-new/rootsurfacecontainer.h [hunk 8]
```diff
@@ -94,6 +94,7 @@ public slots:
 
     WOutputLayout *m_outputLayout = nullptr;
     OutputListModel *m_outputModel = nullptr;
+    Output * m_currentOutput = nullptr;
```
Comment (zccrs) [inline examples/tinywl-new/rootsurfacecontainer.h:97]: Remove

**Review by zccrs — CHANGES_REQUESTED** [review]

File: examples/tinywl-new/OutputMenuBar.qml (hunk 1 above)
Comment (Dami-star) [reply examples/tinywl-new/OutputMenuBar.qml]: done

File: examples/tinywl-new/helper.cpp (hunk 2 above)
Comment (Dami-star) [reply examples/tinywl-new/helper.cpp]: done

File: examples/tinywl-new/helper.cpp (hunk 3 above)
Comment (Dami-star) [reply examples/tinywl-new/helper.cpp]: done

File: examples/tinywl-new/helper.cpp (hunk 4 above)
Comment (Dami-star) [reply examples/tinywl-new/helper.cpp]: done

File: examples/tinywl-new/helper.cpp (hunk 5 above)
Comment (Dami-star) [reply examples/tinywl-new/helper.cpp]: done

File: examples/tinywl-new/output.cpp (hunk 6 above)
Comment (Dami-star) [reply examples/tinywl-new/output.cpp]: done

File: examples/tinywl-new/rootsurfacecontainer.cpp (hunk 7 above)
Comment (Dami-star) [reply examples/tinywl-new/rootsurfacecontainer.cpp]: done

File: examples/tinywl-new/rootsurfacecontainer.h (hunk 8 above)
Comment (Dami-star) [reply examples/tinywl-new/rootsurfacecontainer.h]: done

File: examples/tinywl-new/helper.h [hunk 9]
```diff
@@ -64,13 +64,20 @@ class Helper : public WSeatEventFilter
...
     Q_PROPERTY(float animationSpeed READ animationSpeed WRITE setAnimationSpeed NOTIFY animationSpeedChanged FINAL)
+    Q_PROPERTY(ShowMode outputMode READ outputMode WRITE setOutputMode NOTIFY outputModeChanged FINAL)
     QML_ELEMENT
     QML_SINGLETON
 
 public:
     explicit Helper(QObject *parent = nullptr);
     ~Helper();
 
+    enum class ShowMode {
```
Comment (zccrs) [inline examples/tinywl-new/helper.h:75]: `OutputMode`

**Review by zccrs — CHANGES_REQUESTED** [review]

File: examples/tinywl-new/helper.cpp [hunk 10]
```diff
@@ -644,6 +652,62 @@ Output *Helper::getOutput(WOutput *output) const
     return nullptr;
 }
 
+bool Helper::addOutput()
+{
+    qw_output *newOutput = nullptr;
+    qobject_cast<qw_multi_backend*>(m_backend->handle())->for_each_backend([] (wlr_backend *backend, void *) {
+        qw_output *newOutput = nullptr;
```
Comment (wineee) [inline examples/tinywl-new/helper.cpp:659]:
```suggestion
```

File: examples/tinywl-new/helper.h (hunk 9 above)
Comment (Dami-star) [reply examples/tinywl-new/helper.h]: done

File: examples/tinywl-new/helper.cpp (hunk 10 above)
Comment (Dami-star) [reply examples/tinywl-new/helper.cpp]: done

File: examples/tinywl-new/CopyOutput.qml [hunk 11]
```diff
@@ -8,7 +8,7 @@ OutputItem {
     id: outputItem
 
     required property OutputItem targetOutputItem
```
Comment (zccrs) [inline examples/tinywl-new/CopyOutput.qml:10]: `OutputItem` => `PrimaryOutput`

File: examples/tinywl-new/helper.cpp [hunk 12]
```diff
@@ -158,13 +158,18 @@ void Helper::init()
     auto wOutputManager = m_server->attach<WOutputManagerV1>();
     connect(m_backend, &WBackend::outputAdded, this, [this, wOutputManager] (WOutput *output) {
         allowNonDrmOutputAutoChangeMode(output);
-        auto o = Output::createPrimary(output, qmlEngine(), this);
-        o->outputItem()->setParentItem(m_renderWindow->contentItem());
-        o->outputItem()->stackBefore(m_surfaceContainer);
+        Output *o;
+        if (m_mode == OutputMode::Extension) {
```
Comment (zccrs) [inline examples/tinywl-new/helper.cpp:162]: || `!m_surfaceContainer->primaryOutput()`

File: examples/tinywl-new/helper.cpp [hunk 13]
```diff
@@ -644,6 +646,44 @@ Output *Helper::getOutput(WOutput *output) const
...
+    for (int i = 0; i < m_outputList.size(); i++) {
+        if (m_outputList.at(i) == m_surfaceContainer->primaryOutput())
+            continue;
+
+        Output *o;
+        if (mode == OutputMode::Copy) {
+            o = Output::createCopy(m_outputList.at(i)->output(), m_surfaceContainer->primaryOutput(), qmlEngine(), this);
+            m_surfaceContainer->removeOutput(m_outputList.at(i));
+        } else if(mode == OutputMode::Extension) {
+            o = Output::createPrimary(m_outputList.at(i)->output(), qmlEngine(), this);
```
Comment (zccrs) [inline examples/tinywl-new/helper.cpp:676]: Don't re-create if the outout is primary output

**Review by zccrs — CHANGES_REQUESTED** [review]

File: examples/tinywl-new/CopyOutput.qml (hunk 11 above)
Comment (Dami-star) [reply examples/tinywl-new/CopyOutput.qml:10]: done

File: examples/tinywl-new/helper.cpp (hunk 12 above)
Comment (Dami-star) [reply examples/tinywl-new/helper.cpp]: done

File: examples/tinywl-new/helper.cpp [hunk 14]
```diff
@@ -158,13 +158,18 @@ void Helper::init()
...
         allowNonDrmOutputAutoChangeMode(output);
-        auto o = Output::createPrimary(output, qmlEngine(), this);
-        o->outputItem()->setParentItem(m_renderWindow->contentItem());
-        o->outputItem()->stackBefore(m_surfaceContainer);
+        Output *o;
+        if (m_mode == OutputMode::Extension || !m_surfaceContainer->primaryOutput()) {
+            o = Output::createPrimary(output, qmlEngine(), this);
+            o->outputItem()->stackBefore(m_surfaceContainer);
+            m_surfaceContainer->addOutput(o);
+        } else if (m_mode == OutputMode::Copy && m_surfaceContainer->primaryOutput()) {
```
Comment (zccrs) [inline examples/tinywl-new/helper.cpp:166]: Not needs check primaryOutput here

File: examples/tinywl-new/output.cpp [hunk 15]
```diff
@@ -61,9 +69,20 @@ Output *Output::createCopy(WOutput *output, Output *proxy, QQmlEngine *engine, Q
     o->m_proxy = proxy;
     obj->setParent(o);
 
+    auto contentItem = Helper::instance()->window()->contentItem();
+    outputItem->setParentItem(contentItem);
+    o->updatePrimaryOutputHardwareLayers();
+    connect(getOnscreenViewport(proxy), &WOutputViewport::hardwareLayersChanged,
```
Comment (zccrs) [inline examples/tinywl-new/output.cpp:75]: Where is re-connect if the on screen viewport changed

File: examples/tinywl-new/output.cpp [hunk 16]
```diff
@@ -371,6 +395,45 @@ void Output::layoutAllSurfaces()
...
+    Q_ASSERT(layout);
+
+    auto *layoutOutput = layout->get(output()->nativeHandle());
+    QPointF pos(layoutOutput->x, layoutOutput->y);
+    m_item->setPosition(pos);
+}
+
+std::pair<WOutputViewport*, QQuickItem*> Output::getOutputItemProperty()
+{
+    WOutputViewport *viewportCopy = outputItem()->findChild<WOutputViewport*>({}, Qt::FindDirectChildrenOnly);
```
Comment (zccrs) [inline examples/tinywl-new/output.cpp:410]: Why not use getOnscreenViewport

File: examples/tinywl-new/output.h [hunk 17]
```diff
@@ -55,11 +60,16 @@ class Output : public SurfaceListModel
     QRectF geometry() const;
     QRectF validRect() const;
     QRectF validGeometry() const;
+    std::pair<WOutputViewport*, QQuickItem*> getOutputItemProperty();
```
Comment (zccrs) [inline examples/tinywl-new/output.h:63]: To private

File: examples/tinywl-new/output.h [hunk 18]
```diff
@@ -85,6 +95,8 @@ class Output : public SurfaceListModel
     QList<std::pair<QObject*, int>> m_rightExclusiveZones;
 
     QSizeF m_lastSizeOnLayoutNonLayerSurfaces;
+    QList<WOutputLayer *> m_hardwareLayersOfPrimaryOutput;
+
```
Comment (zccrs) [inline examples/tinywl-new/output.h:99]: ??

**Review by zccrs — CHANGES_REQUESTED** [review]

**Pull request merged**